feat(explore): Show confirmation modal if user exits Explore without saving changes

kgabryje opened the pull request:
### SUMMARY
This PR adds a browser confirmation modal (similar to the one that we have in dashboard edit mode) when user makes a change in control panel and then tries to close Explore (either by closing the tab or moving to another page) without clicking Save button.
Currently, that modal is not customizable. However, when we implement Single Page Application in Superset, we'll probably be able to customize the modal when navigating within Superset.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/15073128/167422541-da6709f4-605b-4740-8cb5-80f693b956ea.png">


### TESTING INSTRUCTIONS
Changes made without saving:
1. Open a chart
2. Make some changes in control panel
3. Try to leave the page (close tab or navigate to another page)
4. Verify that browser confirmation modal appears

Changes made and saved:
1. Open a chart
2. Make some changes in control panel and save.
3. Leave the page, verify that the modal didn't appear

Changes made and then reverted:
1. Open a chart
2. Make some changes in control panel.
3. Revert those changes (for example add and remove a metric)
4. Leave the page, verify that the modal didn't appear

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

kgabryje committed:
feat(explore): Show confirmation modal if user exits Explore without saving changes

michael-s-molina commented:
@kgabryje If I do more than one change, then the confirmation is not appearing. The `beforeunload` is being unregistered.

https://user-images.githubusercontent.com/70410625/167429669-62e98c72-3812-485a-b74a-74d1795c4b8f.mov

kgabryje committed:
Fix calling cleanup func unnecessarily

kgabryje committed:
Fix comparing AdhocMetric instance with JSON object

kgabryje commented:
Well spotted @michael-s-molina! The problem was that the cleanup function in use effect was being called on each change of its dependencies. I moved the cleanup to a separate useEffect that gets called only on unmount.

Also I fixed an old bug that caused the "Altered" pill to stay visible when there was an adhoc metric added and then you added some metric and deleted it immediately (so effectively there was no change, but "Altered" stayed). The reason for that was that in initial form data the adhoc metric is represented by a  plain JSON object, but later it's transformed to AdhocMetric instance, so `isEqual` is false even if the field values are the same. I fixed it by omitting `__proto__` before calling `isEqual` so that we ignore the difference between plain object and class instance

kgabryje commented:
/testenv up

michael-s-molina commented:
Found two more problems:
1 - It's showing the confirmation when navigating using the main header without any changes.
2 - It's showing the confirmation when dropping a filter and canceling

https://user-images.githubusercontent.com/70410625/167842795-3dfbb17e-60ab-4919-b4bc-1626e14234db.mov

kgabryje commented:
> Found two more problems: 1 - It's showing the confirmation when navigating using the main header without any changes. 2 - It's showing the confirmation when dropping a filter and canceling
> 
>  Screen.Recording.2022-05-11.at.8.47.13.AM.mov

Thanks for more testing 🙂
1. It looks like the algorithm for calculating diffs for `Altered` table is faulty, because the confirmation modal uses the same logic. `Altered` pill probably shouldn't popup here as you just opened a saved chart, so the problem might be related. Will investigate.
2. I don't think it's related to dropping and cancelling filter. The `Altered` pill is visible and it's most likely that that's why the confirmation modal is showing.

kgabryje committed:
Replace sliceFormData with the initial form data

kgabryje commented:
/testenv up

michael-s-molina reviewed (approved):
LGTM
